File: 프로그래머스/src/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
int maxExplored = 0;
// https://school.programmers.co.kr/learn/courses/30/lessons/87946# 
/** 모든 경우의 수를 구한다.*/
void dfs(int hp, int explored, vector<bool>& visited, vector<vector<int>>& dungeons) {
    // 현재까지 탐험한 던전 수를 업데이트
    maxExplored = max(maxExplored, explored);

    for (int i = 0; i < dungeons.size(); i++) { // 던전 방문
        if (visited[i]) continue; // 이미 방문한 던전은 건너뜀
        if (hp < dungeons[i][0]) continue; // 최소 피로도가 부족한 경우 건너뜀
        // 소모 필요도는 계산할 필요 없네... 최소 피로도가 소모 피로도 보다 높으니
        
        // 던전을 탐험
        visited[i] = true;
        dfs(hp - dungeons[i][1], explored + 1, visited, dungeons);
        visited[i] = false; // 백트래킹: 탐험 후 다시 방문하지 않은 상태로 되돌림
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    int answer = 0; // count
    int hp = k;
    vector<bool> visited(dungeons.size(),false);
    dfs(hp,0, visited, dungeons);
    
    return maxExplored;
}

File: 2026/완전탐색/피로도.py
# https://school.programmers.co.kr/learn/courses/30/lessons/87946?language=python3
from itertools import *
# queue append pop(0)
# k : 현재 피로도.dungeons[최소필요피로도, 소모필요도]
def solution(k, dungeons):
    answer = -1
    
    def bfs(k, dungeons):
        # contains index of the dungeon
        visited = [False] * len(dungeons)
        queue = [0] # just start with the first one
        hp = k
        local_max = 0
        
        # visit first place
        idx = 0
        
        while(queue):
            # print("======================")
            idx = queue[-1]
            queue.pop(0)
            # print("current index : ", idx)
            # print("HP Left : ",hp)
            
            # boundary
            if (idx == len(dungeons)):
                # print("reached end")
                break;
                
            # already visited
            if(visited[idx]):
                # print("visited.")
                continue
            
            # has min hp and can visit
            if (hp >= dungeons[idx][0] and hp-dungeons[idx][1]>=0):
                # print("HP Left : ", hp, " min : ", dungeons[idx][0], " req : ",dungeons[idx][1])
                visited[idx] = True
                queue.append(idx+1)
                hp -= dungeons[idx][1]
                local_max += 1
             
        # print("local max?", local_max)
        return local_max
        
    # for combinations of len(dungeons)
    # answer = bfs(k)
    for x in permutations(dungeons, len(dungeons)):
        # answer = max(answer, bfs(k))
        temp = bfs(k, x)
        answer = max(answer, temp)
    return answer